xds/src/main/java/io/grpc/xds/EdsLoadBalancer.java

```diff
@@ -0,0 +1,715 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.CONNECTING;
+import static io.grpc.ConnectivityState.IDLE;
+import static io.grpc.ConnectivityState.READY;
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.grpc.ChannelLogger;
+import io.grpc.ChannelLogger.ChannelLogLevel;
+import io.grpc.ConnectivityState;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancerProvider;
+import io.grpc.LoadBalancerRegistry;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import io.grpc.util.ForwardingLoadBalancerHelper;
+import io.grpc.xds.ClientLoadCounter.LoadRecordingSubchannelPicker;
+import io.grpc.xds.ClientLoadCounter.MetricsObservingSubchannelPicker;
+import io.grpc.xds.ClientLoadCounter.MetricsRecordingListener;
+import io.grpc.xds.EnvoyProtoData.DropOverload;
+import io.grpc.xds.EnvoyProtoData.LbEndpoint;
+import io.grpc.xds.EnvoyProtoData.Locality;
+import io.grpc.xds.EnvoyProtoData.LocalityLbEndpoints;
+import io.grpc.xds.InterLocalityPicker.WeightedChildPicker;
+import io.grpc.xds.OrcaOobUtil.OrcaReportingConfig;
+import io.grpc.xds.OrcaOobUtil.OrcaReportingHelperWrapper;
+import io.grpc.xds.XdsClient.EndpointUpdate;
+import io.grpc.xds.XdsClient.EndpointWatcher;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+
+/**
+ * Load balancer for experimental_eds LB policy.
+ */
+final class EdsLoadBalancer extends LoadBalancer {
+
+  private final Helper helper;
+  private final XdsClient xdsClient;
+  private final ChannelLogger channelLogger;
+
+  private final EndpointWatcherImpl.PickerFactory pickerFactory;
+  private final LoadBalancerRegistry lbRegistry;
+  private final ThreadSafeRandom random;
+  private final LoadStatsStore loadStatsStore;
+  private final OrcaPerRequestUtil orcaPerRequestUtil;
+  private final OrcaOobUtil orcaOobUtil;
+
+
+  @Nullable
+  private String edsServiceName;
+
+  @CheckForNull
+  private EndpointWatcherImpl endpointWatcher;
+
+  // TODO(zdapeng): Pass in a nullable LoadStatsStore for LRS.
+  EdsLoadBalancer(Helper helper, XdsClient xdsClient) {
```

Comment:
Why `XdsClient` is taken in constructor? It should be looked up in `ResolvedAddresses`. If a reference is cannot be found, construct one.

Reply:
EDS balancer is a child balancer it dose not need to look up `XdsClient` from `ResolvedAddresses`, it can be passed from its parent balancer. Only `XdsLoadBalancer` need to lookup `XdsClient` from `ResolvedAddresses`, because `XdsLoadBalancer` is created from `LoadBalancerProvider`, the only way to get reference to `XdsClient` is from Attributes in `ResolvedAddresses` or if no found construct one.

Reply:
Why would there be an `XdsLoadBalancer`? The channel should create `CdsLoadBalancer`/`EdsLoadBalancer` directly based on the service config.

Reply:
>Why would there be an XdsLoadBalancer? 

The xds name resolver resolves xds:/// address and provides a service config with lb config, the lb config specifies "xds" policy or "experimental_xds" (Similarly the internal prod name resolver also resolves address and provides a service config with "xds" policy), then `AutoConfiguredLoadBalancerFactory` will then create a balancer for "xds" policy based on the policy _name_, that is `XdsLoadBlancer`.  `CdsLoadBalancer` and `EdsLoadBalancer` are different, they are not created by  `AutoConfiguredLoadBalancerFactory` based on the lb policy _name_ in the service config from name resolver.

Reply:
In the new design, the name resolver will not give "experimental_xds", it will give "experimental_cds" or "experimental_weighted_target" (TBD).

Reply:
That is still not for EDS. For the first release it will be "experimental_cds", so there will be `CdsLoadBalancer` that looks up `XdsClient` from name resolver attributes. `EdsLoadBalancer` does not interact with name resolver directly.

Reply:
We are doing this step-by-step. At this stage, we do not have CDS, so we will make name resolver give "experimental_eds", which lets `AutoConfiguredLoadBalancerFactory` create an `EdsLoadBalancer`.

Look at C++'s implementation. https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/lb_policy/xds/xds.cc

Reply:
Name resolver giving "experimental_eds" is not a usecase, why do we support it? It needs to add an `EdsLoadBalancerProvider` class, and needs to look up `XdsClient` from name resolver attributes and create one if not available, all these are not going to happen in production or integration tests. It's over-engineering.

Reply:
> needs to look up XdsClient from name resolver attributes and create one if not available

https://github.com/grpc/grpc/blob/438cfa71144510fdd44f7702665d0b2c9f8c48ce/src/core/ext/filters/client_channel/lb_policy/xds/xds.cc#L419-L421

Reply:
Regarding "experimental_eds", we are still using "xds_experimental", but  `AutoConfiguredLoadBalancerFactory` creates an `EdsLoadBalancer`. After we add CDS, we will be using "experimental_cds".

Reply:
Edit: 

This is about implementation. There is another design, which is closer to C++'s.

You make an `XdsLoadBalancer` (extends `LoadBalancer`), which encapsulates `CdsPolicy` and `EdsPolicy` (__they do not extend `LoadBalancer`__). `AutoConfiguredLoadBalancerFactory` creates an `XdsLoadBalancer`. You try to find the reference of `XdsClient` in `handleResolvedAddresses` method. If not found, create one. This `XdsClient` is shared by both `CdsPolicy` and `EdsPolicy`.

Reply:
Another thought: From the spec only `CdsLoadBalancer` (and `WeightedTargetLoadBalancer` in the future) need `XdsClient` from `XdsNameResolver` via `ResolvedAddresses` attributes, if they are not found we should fail (We should not support usecases that do not exist). All other balancers can use `XdsClient` from constructor directly (We may ref count it if it is shared).

Reply:
I believe the reason to let LB policies create their own `XdsClient` instance is to allow using xDS load balancer without xDS resolver. So all LB policies should either get the `XdsClient` from `ResolvedAddresses` attributes or create one for themselves. Either way, the point is, all LB policies should share the same `XdsClient` instance.

Reply:
Yes, all lbs share the same `XdsClient` instance.